### PR TITLE
Disconnect function doesn't exists in types of currentProvider

### PR DESCRIPTION
**Pull request by avkos**
## Description

#5864

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist for 1.x:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` with success.
- [ ] I have tested the built `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

## Checklist for 4.x:

- [ ] I have selected the correct 4.x base branch.
- [ ] Within the description, the feature or issue is discussed in detail or an issue is linked.   
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added any required tests for the changes I made 
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `yarn` successfully
- [ ] I ran `yarn lint` successfully
- [ ] I ran `yarn build:web` successfully
- [ ] I ran `yarn test:unit` successfully
- [ ] I ran `yarn test:integration` successfully
- [ ] I ran `compile:contracts` successfully
- [ ] I have tested my code.
- [ ] I have updated the corresponding `CHANGELOG.md` file in the packages I have edited.

### CODE REVIEW

File: packages/web3-utils/src/socket_provider.ts [hunk 1]
```diff
@@ -216,11 +216,12 @@ export abstract class SocketProvider<
 	 * @param data - The data to be sent to the server
 	 */
 	public disconnect(code?: number, data?: string): void {
+		const disconnectCode = code ?? 1000;
```
Comment (Muhammad-Altabba) [inline packages/web3-utils/src/socket_provider.ts]: Where this 1000 came from? Could we please reference a resource to its origin?
And I think it is ideally to put this 1000 in a constant and export it from `web3-types`.

**Review by Muhammad-Altabba — CHANGES_REQUESTED** [review]
I see that `web3.currentProvider?.disconnect()` still gives an error, right? And this is because at web3-core the Web3Context.currentProvider wrongly returns `string` and it has to be converted from 
```ts
public get currentProvider(): SupportedProviders<API> | string | undefined {
    return this.requestManager.provider;
}
```
To:
```ts
public get currentProvider(): SupportedProviders<API> | undefined {
    return this.requestManager.provider;
}
```
Or what do you think?

File: packages/web3-utils/src/socket_provider.ts (hunk 1 above)
Comment (avkos) [reply packages/web3-utils/src/socket_provider.ts]: ok I will add this as a comment https://github.com/Luka967/websocket-close-codes

**Comment by avkos** [comment]
> I see that `web3.currentProvider?.disconnect()` still gives an error, right? And this is because at web3-core the Web3Context.currentProvider wrongly returns `string` and it has to be converted from
> 
> ```ts
> public get currentProvider(): SupportedProviders<API> | string | undefined {
>     return this.requestManager.provider;
> }
> ```
> 
> To:
> 
> ```ts
> public get currentProvider(): SupportedProviders<API> | undefined {
>     return this.requestManager.provider;
> }
> ```
> 
> Or what do you think?

yes. this is first that I tried. but it doesn't help. if we remove `string` currentProvider somehow related to legacy provider. So  I think we need to have conditional type here

**Review by jdevcs — COMMENTED** [review]
pls can you also add a test for `web3.currentProvider?.disconnect();`

File: packages/web3/test/integration/web3.test.ts [hunk 2]
```diff
@@ -75,6 +75,16 @@ describe('Web3 instance', () => {
 		expect(() => new Web3()).not.toThrow();
 	});
 
+	it('check disconnect function', async () => {
```
Comment (luu-alex) [inline packages/web3/test/integration/web3.test.ts:82]: can you test other `supportedproviders` on top of `clienturl` and check functions exist 👍

Comment (avkos) [reply packages/web3/test/integration/web3.test.ts:82]: done

File: packages/web3-providers-ws/test/fixtures/proxy.ts [hunk 3]
```diff
@@ -48,14 +48,26 @@ export const createProxy = async (
 		});
 		closeFunc = async () => {
 			await new Promise(resolve => {
+				const timeOut = setTimeout(() => {
+					// eslint-disable-next-line no-console
+					console.log('timeout1');
```
Comment (luu-alex) [inline packages/web3-providers-ws/test/fixtures/proxy.ts]: remove console.log

File: packages/web3-providers-ws/test/fixtures/proxy.ts [hunk 4]
```diff
@@ -48,14 +48,26 @@ export const createProxy = async (
...
 					ws.removeAllListeners();
+					clearTimeout(timeOut);
 					resolve(true);
 				});
 				ws.terminate();
 			});
 			await new Promise(resolve => {
+				const timeOut = setTimeout(() => {
+					// eslint-disable-next-line no-console
+					console.log('timeout2');
```
Comment (luu-alex) [inline packages/web3-providers-ws/test/fixtures/proxy.ts]: remove debug logs

**Review by luu-alex — APPROVED** [review]

**Comment by luu-alex** [comment]
looks good to me, just remove console logs

File: packages/web3-providers-http/src/index.ts [hunk 5]
```diff
@@ -112,7 +112,7 @@ export default class HttpProvider<
 
 	/* eslint-disable class-methods-use-this */
 	public disconnect() {
-		throw new MethodNotImplementedError();
+		// NO OP
```
Comment (Muhammad-Altabba) [inline packages/web3-providers-http/src/index.ts]: Why to stop throwing on only this specific method on HttpProvider?
I think we should keep this method as it was, similar to other methods in this class.

**Review by jdevcs — APPROVED** [review]
a part from changes required ( existing feedback from other members ) , lgtm .

File: packages/web3-providers-http/src/index.ts (hunk 5 above)
Comment (jdevcs) [reply packages/web3-providers-http/src/index.ts]: we should throw error in http instead of no action for disconnect.

Comment (avkos) [reply packages/web3-providers-http/src/index.ts]: ok

Comment (avkos) [reply packages/web3-providers-http/src/index.ts]: done

**Review by Muhammad-Altabba — APPROVED** [review]
Thanks,

**Pull request merged**